marlowe-runtime/cli/Language/Marlowe/Runtime/CLI/Command/Apply.hs

```diff
@@ -0,0 +1,197 @@
+{-# LANGUAGE ApplicativeDo #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ExistentialQuantification #-}
+
+module Language.Marlowe.Runtime.CLI.Command.Apply
+  where
+
+import Data.ByteString (ByteString)
+import qualified Data.Text as T
+import Language.Marlowe (POSIXTime(..))
+import qualified Language.Marlowe.Core.V1.Semantics.Types as V1
+import qualified Language.Marlowe.Core.V1.Semantics.Types.Address as V1
+import Language.Marlowe.Runtime.CLI.Command.Tx (TxCommand, txCommandParser)
+import Language.Marlowe.Runtime.CLI.Monad (CLI)
+import Language.Marlowe.Runtime.CLI.Option (txOutRefParser)
+import Language.Marlowe.Runtime.ChainSync.Api (unPolicyId)
+import Language.Marlowe.Runtime.Core.Api (ContractId(..), IsMarloweVersion(..), MarloweVersionTag(..))
+import qualified Language.Marlowe.Util as V1
+import Options.Applicative
+import qualified Plutus.V1.Ledger.Api as P
+
+data ApplyCommand = V1ApplyCommand
+  { contractId :: ContractId
+  , inputs :: ContractInputs 'V1
+  , validityLowerBound :: Maybe POSIXTime
+  , validityUpperBound :: Maybe POSIXTime
+  }
+
+data ContractInputs v
+  = ContractInputsByFile FilePath
+  | ContractInputsByValue (Redeemer v)
+  | ContractInputsByValueWithContinuations (Redeemer v) [FilePath]
+
+applyCommandParser :: ParserInfo (TxCommand ApplyCommand)
+applyCommandParser = info (txCommandParser parser) $ progDesc "Apply inputs to a contract"
+  where
+    parser = V1ApplyCommand
```

Comment:
So if I want to apply several inputs at once (e.g., `choose`, `deposit`, `choose`), I would need to use the `apply` command with an input JSON file, but if I just want to apply one input I can use `choose`, `deposit`, or `notify`.

If I want to trigger a timeout (i.e., no input at all), I'd need to use `apply` with an input file containing an empty JSON array. We could make `--input-file` optional so that one wouldn't need to create an empty file just to trigger a timeout.

A more unified approach would be to just have `apply`:
1. Either no input-related arguments, which would send empty input to trigger a timeout,
2. or `--input-file` to let the user send JSON with an array of whatever input they want,
3. or something like a sequence of `--deposit`, `--choose`, `--notify` along with the specifics for each of the inputs.

Reply:
That is a good point, I forgot about empty inputs. We could also add an "exec" command which was a shortcut for "apply" with no inputs.